bitcoin/src/blockdata/script/push_bytes.rs

```diff
@@ -0,0 +1,463 @@
+//! Contains `PushBytes` & co
+
+use core::ops::{Deref, DerefMut};
+use core::borrow::{Borrow, BorrowMut};
+#[allow(unused)]
+use crate::prelude::*;
+
+pub use primitive::*;
+
+/// This module only contains required operations so that outside functions wouldn't accidentally
+/// break invariants. Therefore auditing this module should be sufficient.
+mod primitive {
+    #[allow(unused)]
+    use crate::prelude::*;
+
+    use super::PushBytesError;
+    use core::convert::{TryFrom, TryInto};
+    use core::ops::{Index, Range, RangeFull, RangeFrom, RangeTo, RangeInclusive, RangeToInclusive};
+    #[cfg(feature = "rust_v_1_53")]
+    use core::ops::Bound;
+
+    #[cfg(any(target_pointer_width = "16", target_pointer_width = "32"))]
+    fn check_limit(len: usize) -> Result<(), PushBytesError> {
+        Ok(())
+    }
+
+    #[cfg(not(any(target_pointer_width = "16", target_pointer_width = "32")))]
+    fn check_limit(len: usize) -> Result<(), PushBytesError> {
+        if len < 0x100000000 {
+            Ok(())
+        } else {
+            Err(PushBytesError { len })
+        }
+    }
+
+    /// Byte slices that can be in Bitcoin script.
+    ///
+    /// The encoding of Bitcoin script restricts data pushes to be less than 2^32 bytes long.
+    /// This type represents slices that are guaranteed to be within the limit so they can be put in
+    /// the script safely.
+    #[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    #[repr(transparent)]
+    pub struct PushBytes([u8]);
+
+    impl PushBytes {
+        /// Creates `&Self` without checking the length.
+        ///
+        /// ## Safety
+        ///
+        /// The caller is responsible for checking that the length is less than the [`LIMIT`].
+        unsafe fn from_slice_unchecked(bytes: &[u8]) -> &Self {
+            &*(bytes as *const [u8] as *const PushBytes)
+        }
+
+        /// Creates `&mut Self` without checking the length.
+        ///
+        /// ## Safety
+        ///
+        /// The caller is responsible for checking that the length is less than the [`LIMIT`].
+        unsafe fn from_mut_slice_unchecked(bytes: &mut [u8]) -> &mut Self {
+            &mut *(bytes as *mut [u8] as *mut PushBytes)
+        }
+
+        /// Creates an empty `PushBytes`.
+        pub fn empty() -> &'static Self {
+            // 0 < LIMIT
+            unsafe { Self::from_slice_unchecked(&[]) }
+        }
+
+        /// Returns the underlying bytes.
+        pub fn as_bytes(&self) -> &[u8] {
+            &self.0
+        }
+
+        /// Returns the underlying mutbale bytes.
+        pub fn as_mut_bytes(&mut self) -> &mut [u8] {
+            &mut self.0
+        }
+    }
+
+    macro_rules! delegate_index {
```

Comment:
In bcd95fa0368c12c3cc8e01bdc896cc738054b648:

I think this macro can be replaced by a suitable generic `impl Index<I> for PushBytes where Vec: Index<I>, Vec: Index<I>::Output = [u8]` or something.

Should be done in a followup, I don't want to have to re-review this thing and it's not a big deal anyay :)

Reply:
Then safe code could break our invariant:

```rust
strut Evil;

impl Index<Evil> for [u8] {
    type Output = [u8];

    fn index(&self, index: Evil) -> &Self::Output {
        static BIG_ARR: [u8; 4294967297] = [0; 4294967297];
        &BIG_ARR
    }
}

script.push(&PushBytes::empty()[Evil]);
```

Reply:
Lol! Ok, that's fair.